File: Algorithms/142. Linked List Cycle II/Solution.py
#coding=utf-8
__author__ = 'xuxuan'
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def detectCycle(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        fast=head
        slow=head
        while fast is not None and fast.next is not None:
            fast=fast.next.next
            slow=slow.next
            if slow==fast:
                fast=head
                while fast is not slow:
                    fast=fast.next
                    slow=slow.next
                return fast
        return None

File: Algorithms/142. Linked List Cycle II/Solution.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode detectCycle(ListNode head) {
    	ListNode fast=head;
    	ListNode slow=head;
        while (true) {
			if (fast==null||fast.next==null) {
				return null;
			}
			fast=fast.next.next;
			slow=slow.next;
			if (fast==slow) {
				ListNode again=head;
				while (true) {
					if (slow==again) {
						return again;
					}else {
						slow=slow.next;
						again=again.next;
					}
				}
			}
		}
    }
}